### PR TITLE
!feat: caddy@2

### LINKED ISSUE #35: update to caddy 2.0.0
With caddy being in version 2.0.0 and caddy-gen still sporting version 0.10.12 wouldn't it be a good idea to upgrade to the nwest version?

I tried setting the caddy version in the dockerfile myself, but it wont run. I suppose there is some new stuff about the api in the Caddyfile. 

But maybe there is no real reasson for an update?

### PR DESCRIPTION
**Pull request by Siemko**
A lot of breaking changes, but not so scary... Closes #35.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,46 +1,28 @@
...
-ARG FOREGO_VERSION="0.16.1"
 ARG DOCKER_GEN_VERSION="0.7.4"
+ARG FOREGO_VERSION="0.16.1"
+
 
 ENV CADDYPATH="/etc/caddy"
 ENV DOCKER_HOST="unix:///tmp/docker.sock"
 
-
-# Install all dependenices:
```
Comment (sobolevn) [inline Dockerfile:15]: Let's keep the comments 👍

File: README.md [hunk 2]
```diff
@@ -19,6 +18,26 @@ But when you need to scale your application Caddy is limited to its static confi
 To overcome this issue we are using `docker-gen` to generate configuration everytime a container spawns or dies.
 Now scaling is easy!
 
+## CADDY 2
+
+BREAKING CHANGES HERE!
```
Comment (sobolevn) [inline README.md:23]: Let's add `Breaking change since <v...>` with the link to the previous version.

File: README.md [hunk 3]
```diff
@@ -19,6 +18,26 @@ But when you need to scale your application Caddy is limited to its static confi
...
+- `virtual.auth.username` and
+- `virtual.auth.password` together provide HTTP basic authentication.
+
+Password should be a string `base64` encoded from `bcrypt` hash. You can use https://bcrypt-generator.com/ with default config and https://www.base64encode.org/.
+
+## Backing up certificates
+
+To backup certificates make a volume:
+
+`./caddy-info:/data/caddy`
```
Comment (sobolevn) [inline README.md:40]: Let's add a bit more context here (with yaml like this):

```yml
services:
  caddy:
    volumes:
      - ./caddy-info:/data/caddy
```

File: README.md [hunk 4]
```diff
@@ -32,30 +51,29 @@ services:
     image: "wemakeservices/caddy-gen:latest"
     restart: always
     volumes:
-      - /var/run/docker.sock:/tmp/docker.sock:ro  # needs socket to read events
-      - ./certs/acme:/etc/caddy/acme  # to save acme
-      - ./certs/ocsp:/etc/caddy/ocsp  # to save certificates
+      - /var/run/docker.sock:/tmp/docker.sock:ro # needs socket to read events
+      - ./certs/acme:/etc/caddy/acme # to save acme
+      - ./certs/ocsp:/etc/caddy/ocsp # to save certificates
```
Comment (sobolevn) [inline README.md:56]: I though that some paths have changed, haven't they?

File: docker-gen/config/docker-gen.cfg [hunk 5]
```diff
@@ -3,4 +3,4 @@ template = "/code/docker-gen/templates/Caddyfile.tmpl"
 dest = "/etc/caddy/Caddyfile"
 onlyexposed = false
 watch = true
-wait = "500ms:2s"
+wait = "500ms:2s"
```
Comment (sobolevn) [inline docker-gen/config/docker-gen.cfg:6]: Let's keep all the newlines, that's what our `.editorconfig` specifies.

Comment (Siemko) [reply docker-gen/config/docker-gen.cfg:6]: :heavy_check_mark:

File: README.md (hunk 4 above)
Comment (Siemko) [reply README.md]: :heavy_check_mark:

File: README.md (hunk 3 above)
Comment (Siemko) [reply README.md]: :heavy_check_mark:

File: README.md (hunk 2 above)
Comment (Siemko) [reply README.md]: :heavy_check_mark:

File: Dockerfile (hunk 1 above)
Comment (Siemko) [reply Dockerfile:15]: :heavy_check_mark:

**Comment by Siemko** [comment]
All right @sobolevn, I've made those suggested changes, please let me know if it's all right with you :)

File: README.md [hunk 6]
```diff
@@ -20,7 +20,7 @@ Now scaling is easy!
 
 ## CADDY 2
 
-BREAKING CHANGES HERE!
+BREAKING CHANGES since [version 0.30.0](https://github.com/wemake-services/caddy-gen/releases/tag/0.3.0)!
```
Comment (sobolevn) [inline README.md:23]: Typo: `0.30.0`, should be `0.3.0`

File: README.md [hunk 7]
```diff
@@ -52,8 +57,7 @@ services:
     restart: always
     volumes:
       - /var/run/docker.sock:/tmp/docker.sock:ro # needs socket to read events
-      - ./certs/acme:/etc/caddy/acme # to save acme
-      - ./certs/ocsp:/etc/caddy/ocsp # to save certificates
+      - ./caddy-info:/data/caddy
```
Comment (sobolevn) [inline README.md:60]: Let's add a comment on why do we need this line

Comment (Siemko) [reply README.md]: :heavy_check_mark:

File: README.md (hunk 6 above)
Comment (Siemko) [reply README.md]: :heavy_check_mark:

**Comment by sobolevn** [comment]
Hey @Siemko! 

Do you need any help with fixing the CI?

**Comment by Siemko** [comment]
Yes @sobolevn, I would be much obliged, I haven't use Travis CI before, and now I have limited time to learn that.

**Comment by sobolevn** [comment]
So, there are two problems right now:

```
$ curl 127.0.0.1:2015
curl: (56) Recv failure: Connection reset by peer
The command "curl 127.0.0.1:2015" exited with 56.
0.48s

$ disl caddy-gen:latest 75MB
caddy-gen:latest exceeds 75MB limit by 10.03 MiB
The command "disl caddy-gen:latest 75MB" exited with 1.
```

The first one: we use to test that Caddy is up and running on port 2015. Now, it does not respond. Since, I am not quite familiar with how `caddy@2` works, it might be just a new configuration issue. Or possibly we have to `sleep` for some time before calling it. What do you think? @Siemko 

The second problem is that the docker image size is now bigger. So, we have to change this line https://github.com/wemake-services/caddy-gen/blob/master/.travis.yml#L19 to be `disl caddy-gen:latest 90MB`

**Comment by Siemko** [comment]
> The first one: we use to test that Caddy is up and running on port 2015. Now, it does not respond. Since, I am not quite familiar with how `caddy@2` works, it might be just a new configuration issue. Or possibly we have to `sleep` for some time before calling it. What do you think? @Siemko

@sobolevn yes, caddy used to listen on port 2015, but now it exposes 80 and 443.

I will try to fix those issues later today or tomorrow.

**Comment by sobolevn** [comment]
Thank you!

**Comment by sobolevn** [comment]
Thanks a lot!

And, @Siemko, welcome on board! 🏴‍☠️

**Comment by AlphaMoonbaseBerlin** [comment]
Thanks alot, works like a charm and feels snappier :)

**Pull request merged**